File: ic-gamejam-2014/src/entity.cpp
#include "main.h"
#include "game.h"
#include "entity.h"

Entity::Entity(const Vec2& p)
    : radius_(20.0f),
      position_(p),
      velocity_(0.0f, 0.0f),
      possessionMeter_(0.0f),
      possessionCooldown_(false)
{
    shape_.setRadius(radius_);
    shape_.setOrigin(Vec2(radius_, radius_));
    shape_.setFillColor(sf::Color(0, 64, 128, 128));
    shape_.setPosition(position_);
}

void Entity::update(ObjectList& objects, float dt)
{
    shared_ptr<Object> possessed = possessed_.lock();

    possessionTimer_ = max(0.0f, min(5.0f, possessionTimer_ + (possessed ? dt : -dt)));

    if (possessed)
    {
        // Add forces to the object
        float force = 400000.0f * dt;
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::W))
        {
            possessed->applyForce(Vec2(0.0f, -force), dt);
        }
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::S))
        {
            possessed->applyForce(Vec2(0.0f, force), dt);
        }
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::A))
        {
            possessed->applyForce(Vec2(-force, 0.0f), dt);
        }
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::D))
        {
            possessed->applyForce(Vec2(force, 0.0f), dt);
        }

        // Disowning the entity
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::Return) || possessionTimer_ > (5.0f - EPSILON))
        {
            position_ = possessed->getPosition();
            velocity_ = Vec2(0.0f, 0.0f);
            possessed_.reset();
            possessionCooldown_ = true;
        }
    }
    else
    {
        float speed = 1500.0f * dt;
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::W))
        {
            velocity_.y -= speed;
        }
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::S))
        {
            velocity_.y += speed;
        }
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::A))
        {
            velocity_.x -= speed;
        }
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::D))
        {
            velocity_.x += speed;
        }

        position_.x += velocity_.x * dt;
        position_.y += velocity_.y * dt;
        shape_.setPosition(position_);

        // Attenuate velocity
        float attenuation = 0.05f;
        velocity_.x = lerp(velocity_.x, 0.0f, attenuation, dt);
        velocity_.y = lerp(velocity_.y, 0.0f, attenuation, dt);

        // If space key is pressed, begin possessing
        float time = 0.25f;
        if (sf::Keyboard::isKeyPressed(sf::Keyboard::Space))
        {
            possessionMeter_ += dt / time;
        }
        else
        {
            possessionMeter_ -= dt / time;
        }
        possessionMeter_ = max(0.0f, min(1.0f, possessionMeter_));
        float scale = sqrt(1.0f - possessionMeter_ * 0.95f);
        shape_.setScale(scale, scale);

        // If the posessionMeter reaches 0, stop cooling down
        if (possessionMeter_ < EPSILON)
        {
            possessionCooldown_ = false;
        }

        // If the possessionMeter reaches 1, take control if not cooling down
        if (possessionMeter_ > (1.0f - EPSILON) && !possessionCooldown_)
        {
            // Search for the object the entity is floating over
            for (auto o : objects)
            {
                Vec2 pos = o->getPosition();
                float distance = pow(pos.x - position_.x, 2) + pow(pos.y - position_.y, 2);
                if (distance < (radius_ * radius_))
                {
                    // Possess this entity
                    possessed_ = o;
                    break;
                }
            }
        }
    }
}

void Entity::render(sf::RenderWindow& window)
{
    shared_ptr<Object> possessed = possessed_.lock();
    if (possessed)
    {
        // Draw some particle effects
    }
    else
    {
        window.draw(shape_);
    }
}

Vec2 Entity::getPosition() const
{
    shared_ptr<Object> possessed = possessed_.lock();
    if (possessed)
    {
        return possessed->getPosition();
    }
    else
    {
        return position_;
    }
}

float Entity::getPossessTimer() const
{
    return 1.0f - possessionTimer_ / 5.0f;
}

weak_ptr<Object> Entity::getPossessedObject()
{
    return possessed_;
}

File: ic-gamejam-2014/rust/src/entity.rs
extern crate rsfml;

use Vec2 = rsfml::system::Vector2f;
use rsfml::window::keyboard;
use rsfml::graphics::{CircleShape, Color};
use std::num::pow;

use std::rc::Rc;

use context;
use object;

pub struct Entity<'e, 'p>
{
  pub position: Vec2,
  velocity: Vec2,

  // Posessed object
  posessed: Option<&object::Object<'p>>,

  // Renderable
  shape: CircleShape<'e>,
}

impl<'e, 'p> Entity<'e, 'p>
{
  // Constructor
  pub fn new(p: Vec2) -> Entity
  {
    // Set up the object
    let mut object = Entity
    {
      position: p,
      velocity: Vec2::new(0.0, 0.0),
      posessed: None,
      shape: match CircleShape::new()
      {
        Some(shape) => shape,
        None => fail!("ERROR: Cannot create the shape")
      }
    };

    object.shape.set_radius(20.0);
    object.shape.set_fill_color(&Color::blue());
    object.shape.set_position(&object.position);

    // Return the object
    object
  }

  // Update
  pub fn update(&mut self, objects: &'p Vec<object::Object<'p>>)
  {
    let speed = 0.001;
    let attenuation = 0.9975;

    if keyboard::is_key_pressed(keyboard::W) { self.velocity.y -= speed; }
    if keyboard::is_key_pressed(keyboard::S) { self.velocity.y += speed; }
    if keyboard::is_key_pressed(keyboard::A) { self.velocity.x -= speed; }
    if keyboard::is_key_pressed(keyboard::D) { self.velocity.x += speed; }

    self.position.x += self.velocity.x;
    self.position.y += self.velocity.y;

    // Attenuate velocity
    self.velocity.x *= attenuation;
    self.velocity.y *= attenuation;

    // Check for collision between this and other objects
    for o in objects.iter()
    {
      let distance = pow(o.position.x - self.position.x, 2) + pow(o.position.y - self.position.y, 2);
      if distance < (32.0 * 32.0)
      {
        self.posessed = Some(o);
        break;
      }
    }
  }

  // Draw
  pub fn draw(&mut self, cxt: &mut context::Context)
  {
    let local_position = self.position - cxt.camera_position;
    self.shape.set_position(&local_position);
    cxt.window.draw(&self.shape);
  }
}
